### PR TITLE
boards: stm32: make flashing with stm32cubeprogrammer available & default on all boards

### LINKED ISSUE #75237: stm32: boards: stm32cubeprogrammer runner support is missing on many ST boards
**Is your enhancement proposal related to a problem? Please describe.**
Flashing with STM32CubeProgrammer runner is not supported on many ST boards

**Describe the solution you'd like**
Flashing with STM32CubeProgrammer runner should be supported on all ST official boards.

**Describe alternatives you've considered**
N/A

**Additional context**
It is just a matter of adding these two lines to the `board.cmake` file
```
board_runner_args(stm32cubeprogrammer "--port=swd" "--reset-mode=hw")

include(${ZEPHYR_BASE}/boards/common/stm32cubeprogrammer.board.cmake)
```

### PR DESCRIPTION
**Pull request by ajarmouni-st**
Add stm32cubeprogrammer where it is not available, & make it the default flashing method. Update doc to list flashing options.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/75237

Edit: to give more context, right now the west runner that's selected when calling `west flash` without `--runner ...`option is kind of arbitrary & depends on which runner is first in `board.cmake`.

### CODE REVIEW

**Comment by erwango** [comment]
> @kartben Do you think this can go in v3.7.0?

I prefer avoiding such bulk change in RC2 context. Doc update is welcome though.

**Comment by glenn-andrews** [comment]
Does this now require people to make an ST account to download STM32CubeProg?

OpenOCD and JLink doesn't need an account on a corporation to download.

**Comment by glenn-andrews** [comment]
> @glenn-andrews it can be downloaded without an account in guest mode, but an email address is needed to receive the link.
> 
> In any case, this PR aims to unify & simplify the user experience when the `--runner` option is not specified. We can put OpenOCD as default, but then Zephyr SDK does not provide a pre-built OpenOCD binary on Windows, & it takes months anyway to add support for the latest STM32 SoCs in the main upstream OpenOCD branch. 

I'd prefer streamlining the installation instructions for OpenOCD. You can't debug with STM32CubeProg, so you need to install something else anyway.

What do we do for debug support until new STM32 SoC are added?

**Comment by JarmouniA** [comment]
> What do we do for debug support until new STM32 SoC are added?

You can debug with ST-link Server (https://www.st.com/en/development-tools/st-link-server.html#overview) or ST-OpenOCD, both are included with STM32CubeCLT (https://www.st.com/en/development-tools/stm32cubeclt.html).

**Comment by glenn-andrews** [comment]
> You can debug with ST-link Server (https://www.st.com/en/development-tools/st-link-server.html#overview) or ST-OpenOCD, both are included with STM32CubeCLT (https://www.st.com/en/development-tools/stm32cubeclt.html).

So should we be documenting to download that package and how to set it up, rather than just the programmer?

**Comment by glenn-andrews** [comment]
> > So should we be documenting to download that package and how to set it up, rather than just the programmer?
> 
> Right now, `west debug` supports only OpenOCD, but I think support should be added for ST-link GDB server as well, esp. since, I believe, there are STM32 targets that can only be debugged with the later, like CM4 core on the dual-core STM32H747I.

How heavy a lift is fixing `west debug`? Right now to use STM32 you install OpenOCD, unless you're working with newer MCUs. After this you'll need to install both OpenOCD and STM32CubeProg to flash and debug (or always specify the OpenOCD runner for flashing).

A one-stop-shop of installing STM32CubeCLT would be preferable than having to install multiple tool packages.

**Review by erwango — DISMISSED** [review]

**Comment by erwango** [comment]
II approved but I forgot the most important: Nice to see you around @JarmouniA :)

File: doc/releases/migration-guide-4.0.rst [hunk 1]
```diff
@@ -27,6 +27,13 @@ Boards
 * :ref:`native_posix<native_posix>` has been deprecated in favour of
   :ref:`native_sim<native_sim>` (:github:`76898`).
 
+STM32
+=====
+
+* On all official STM32 boards, `west flash` selects STM32CubeProgrammer as the default west runner.
+  If you want to enforce the selection of another runner like OpenOCD or pyOCD for flashing, you should
```
Comment (glenn-andrews) [inline doc/releases/migration-guide-4.0.rst:34]: Is there a way in west to permanently select a preferred runner?

i.e. if you _really_ want to not install STM32CubeProg, can you configure west such that it always selects OpenOCD for flashing when no runner is given on the command line?

Comment (erwango) [reply doc/releases/migration-guide-4.0.rst:34]: It is possible using twister and map file.
Using `west`, I thought it was possible using [configuration options](https://docs.zephyrproject.org/latest/develop/west/build-flash-debug.html#configuration-options), which is unfortunately not the case apparently, but I think can be added.

File: boards/st/nucleo_g474re/doc/index.rst [hunk 2]
```diff
@@ -207,6 +207,17 @@ Flashing
...
+The board is configured to be flashed using west `STM32CubeProgrammer`_ runner,
+so its installation is required.
+
+Alternatively, openocd or JLink can also be used to flash the board using
+the ``--runner`` (or ``-r``) option:
+
+.. code-block:: console
+
+   $ west flash --runner openocd
+   $ west flash --runner jlink
```
Comment (str4t0m) [inline boards/st/nucleo_g474re/doc/index.rst:219]: Additionally to stm32cubeprogrammer, it only has pyocd and openocd.
But jlink could be added in a follow up PR.

File: boards/st/stm32f103_mini/doc/index.rst [hunk 3]
```diff
@@ -162,3 +162,6 @@ References
 
 .. _STM32F103 data sheet:
    https://www.st.com/resource/en/datasheet/stm32f103rc.pdf
+
+.. _STM32CubeProgrammer:
+   https://www.st.com/en/development-tools/stm32cubeprog.html
```
Comment (str4t0m) [inline boards/st/stm32f103_mini/doc/index.rst:167]: NIT: I think that link is never referenced.

File: boards/st/nucleo_g431rb/doc/index.rst [hunk 4]
```diff
@@ -193,25 +193,16 @@ Flashing
...
-To manually enable the openocd interface, You can still update, compile and install
-a 'local' openocd from the official openocd repo http://openocd.zylin.com .
-Then run the following openocd command where the '/usr/local/bin/openocd'is your path
-for the freshly installed openocd, given by "$ which openocd" :
-
-.. code-block:: console
-
-   $ west flash --openocd /usr/local/bin/openocd
+   $ west flash --runner openocd
+   $ west flash --runner jlink
```
Comment (str4t0m) [inline boards/st/nucleo_g431rb/doc/index.rst:205]: NIT: Jlink hasn't been added for this board yer.

**Review by str4t0m — DISMISSED** [review]

**Comment by str4t0m** [comment]
Will need a rebase to resolve a conflict in the  migration guide.

**Comment by JarmouniA** [comment]
Rebase

**Comment by JarmouniA** [comment]
@str4t0m your remarks have been addressed.

**Review by erwango — APPROVED** [review]

**Review by str4t0m — APPROVED** [review]

**Pull request merged**